src/core/batch/batch.ts

```diff
@@ -20,47 +23,115 @@ interface BloomFilterData {
   RowGroupIndex: number;
 }
 
+interface BatchFileData {
+  url: URL;
+  hash: HexString;
+}
+
+/**
+ * BatchError indicates that something went wrong in generating a batch file.
+ * This error object will include a fileHandle field containing the un-closed
+ * write stream of the batch.
+ */
+export class BatchError extends Error {
+  fileHandle: WriteStream;
+
+  constructor(message: string, fileHandle: WriteStream) {
+    super(message);
+    this.name = "BatchError";
+    this.fileHandle = fileHandle;
+  }
+}
+
+/**
+ * EmptyBatchError indicates that no messages were passed in attempting to
+ * create a batch file which is not allowed.
+ */
+export class EmptyBatchError extends BatchError {
+  constructor(fileHandle: WriteStream) {
+    super("Invalid message iterator for batch: iterator contains no messages", fileHandle);
+    this.name = "EmptyBatchError";
+  }
+}
+
+/**
+ * MixedDSNPTypeError indicates that more than one type of DSNP message was
+ * passed in attempting to create batch file which is not allowed.
+ */
+export class MixedTypeBatchError extends BatchError {
+  constructor(fileHandle: WriteStream) {
+    super("Invalid message iterator for batch: iterator contains multiple DSNP types", fileHandle);
+    this.name = "MixedTypeBatchError";
+  }
+}
+
+type BatchIterable<T extends DSNPType> = AsyncOrSyncIterable<DSNPMessageSigned<DSNPTypedMessage<T>>>;
+
 /**
- * createFile() takes a series of DSNP messages and returns a URL
- * for storage location.
+ * createFile() takes a series of Batch DSNP messages, writes them to a file at
+ * specified target path and returns a BatchFileData object.
  *
  * @param targetPath - The path to and name of file
+ * @param dsnpType - The DSNPType of the messages provided
  * @param messages - An array of DSNPMessage to include in the batch file
  * @param opts - Optional. Configuration overrides, such as store, if any
- * @returns A URL of the storage location
- * @throws error if messages argument is empty.
+ * @returns A BatchFileData object including a URL and keccak hash of the file
  */
-export const createFile = async (targetPath: string, messages: DSNPMessage[], opts?: ConfigOpts): Promise<URL> => {
-  if (messages.length === 0) throw EmptyArrayError;
-
-  const schema = new ParquetSchema(getSchemaFor(messages[0].dsnpType));
-  const bloomFilterOptions = getBloomFilterOptionsFor(messages[0].dsnpType);
+export const createFile = async <T extends DSNPType>(
+  targetPath: string,
+  dsnpType: T,
+  messages: BatchIterable<T>,
```

Comment:
Due to Enddy's comment below, I think we can also remove this generic and just do:
```suggestion
  messages: BatchIterable<DSNPType>,
```

Why? Because we don't need it to type the return more specifically and the only field we use is `dsnpType`.

Reply:
See earlier reply: https://github.com/LibertyDSNP/sdk-ts/pull/81#discussion_r666197951